From 2f7638663af763c8507e91b8c038286604713e21 Mon Sep 17 00:00:00 2001
From: Varun Kumar Reddy Yaparla <vyaparla@habana.ai>
Date: Sat, 23 Oct 2021 13:12:04 +0300
Subject: [PATCH] [SW-56278] gaudi2: security changes for fake fuse

This change includes support for passing the security
information to later stages by setting the security
boot info read from flash in case of fake security or
from efuse/rom in case of real security

Change-Id: Ifd6d56be7564ae541e171277ff9c8f7544709517
---

diff --git a/arch/gaudi2/common/common.h b/arch/gaudi2/common/common.h
index fd23547..d2e3c70 100644
--- a/arch/gaudi2/common/common.h
+++ b/arch/gaudi2/common/common.h
@@ -8,7 +8,6 @@
 
 #include <preboot.h>
 #include <limits.h>
-#include <gaudi2_blocks_linux_driver.h>
 
 #ifdef TARGET_IS_PLDM
 	#define MME_MASK 0x0
diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index a59cbd3..6388892 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -15,6 +15,10 @@
 #include <uart.h>
 #include <hl_protbits_fw_if.h>
 #include <string.h>
+#include <hl_boot_desc.h>
+#include <boot_info.h>
+#include <spi_flash.h>
+#include <gaudi2/gaudi2_security.h>
 #include <psoc_arc0_mstr_if_core_hbw_regs.h>
 #include <psoc_arc0_mstr_if_core_lbw_regs.h>
 #include <psoc_arc1_mstr_if_core_hbw_regs.h>
@@ -35,6 +39,7 @@
 #include <dcore0_vdec0_mstr_if_core_hbw_regs.h>
 #include <dcore0_vdec0_mstr_if_core_lbw_regs.h>
 #include <pmmu_hbw_mstr_if_core_hbw_regs.h>
+#include "common.h"
 
 static struct hl_sec_rrs_cfg hl_sec_lbw_rrs_tbl[] = {
 	[HL_SEC_ENGINE_IF_RTR] = {
@@ -139,6 +144,10 @@
 		.min = 0,
 		.max = SRAM_END,	/* image end to sram end */
 	},
+};
+
+static struct hl_sec_efuse_cache preboot_hl_sec_fuse_cache = {
+	.is_invalid = true,
 };
 
 static void hl_secure_update_hit_reg(int rr_idx, int inst_dist,
@@ -287,6 +296,66 @@
 		hl_secure_hbw_rr_gen(HL_SEC_HBW_PSOC_LBW, min, max, 0, HL_SEC_PSOC_INST_NUM,
 					HL_SEC_ENGINE_PSOC);
 }
+
+#define SZ_1K			1024
+#define SZ_1M			(SZ_1K * SZ_1K)
+#define FAKE_FUSE_OFFSET	(2 * SZ_1M + 64 * SZ_1K)
+static const struct hl_sec_efuse *preboot_get_hl_sec_fake_efuse(void)
+{
+	struct hl_sec_efuse *fuse;
+	u32 *val;
+	int i;
+
+	if (preboot_hl_sec_fuse_cache.is_invalid) {
+		fuse = &preboot_hl_sec_fuse_cache.fuse;
+		/* Copy efuse security data */
+		spi_flash_copy(&fuse->data, FAKE_FUSE_OFFSET + (EFUSE_SEC_OFFSET * sizeof(*val)),
+					sizeof(*fuse), "FAKE SECURITY FUSE");
+
+		/* Fake fuse data is stored in negative logic */
+		for (i = 0; i < ARRAY_SIZE(fuse->data); i++) {
+			val = &fuse->data[i];
+			*val = ~(*val);
+		}
+
+		preboot_hl_sec_fuse_cache.is_invalid = false;
+	}
+
+	return &preboot_hl_sec_fuse_cache.fuse;
+}
+
+static void preboot_security_sp_sram_write_fake_efuse(void)
+{
+	const struct hl_sec_efuse *efuse;
+
+	efuse = preboot_get_hl_sec_fake_efuse();
+	if (efuse == NULL)
+		return;
+
+	boot_info_set(HL_BOOT_INFO_SECURITY_FUSE, efuse, sizeof(*efuse));
+}
+
+static void preboot_security_sp_sram_write_efuse(void)
+{
+	/* Do nothing for real eFuse */
+}
+
+void preboot_security_sp_sram_write_key_idx(u32 *key_idx)
+{
+	boot_info_set(HL_BOOT_INFO_KEY_IDX, key_idx, sizeof(u32));
+}
+
+static void preboot_security_sp_sram_init(void)
+{
+	if (preboot_fake_security_is_enabled()) {
+		preboot_puts("preboot in fake security mode\n");
+		preboot_security_sp_sram_write_fake_efuse();
+	} else
+		preboot_security_sp_sram_write_efuse();
+}
+
+#define EN_ARC_SEC_PRIV				0x3
+#define AXPROT_OVRD_SEC				0x1
 
 /* protection bits configurations to open some selective registers */
 static u32 gaudi2_protbits_data0[] = { 0x400, 0xc103 };
@@ -717,6 +786,7 @@
 void preboot_security_init(void)
 {
 	preboot_secure_pcie();
+	preboot_security_sp_sram_init();
 
 	/* Do security related initializations only if security is enabled */
 	if (preboot_security_is_enabled()) {
